dspy/teleprompt/finetune_teleprompter.py

```diff
@@ -0,0 +1,237 @@
+
+    # Ensure that the student and teacher programs have the same structure
+    logger.info(_INFO_STRUCTURAL_EQUIVALENCY)
+    student._assert_structural_equivalency(teacher)
+
+    # Ensure that the predictors of the programs point to different objects
+    logger.info(_INFO_SHARED_PREDICTOR)
+    student._assert_no_shared_predictor(teacher)
+
+    # Ensure that the LM consistency property is satisfied
+    logger.info(_INFO_LM_CONSISTENCY)
+    teacher._assert_lm_consistency()
+
+    # If the global LM is being used, set it to the LMs of the copied teacher
+    # program predictors to to avoid handling the same edge cases later
+    if dspy.settings.lm:
+        teacher._set_all_predictor_lms(dspy.settings.lm)
+
+    return teacher
+
+
+# TODO: fix docstring
+def convert_to_module_level_message_data(
+        data: List[Dict],
+        keep_data_keys: bool = False,
+        exclude_demos: bool = False,
+        try_to_record_lm_kwargs: bool = False,
+        program: Program = None
+    ) -> List[Dict]:
+    """Wrapper around the function
+    `build_messages_from_trace`, calling it on the "trace" field
+    of each dictionary in the input data list and combiningin the results into
+    a list of prompt-completion data dictionaries."""
+
+    prompt_completion_data = []
+    for data_dict in data:
+        trace = data_dict["trace"]
+        trace_prompt_comletion_data = build_messages_from_trace(
+            trace=trace, exclude_demos=exclude_demos,
+            try_to_record_lm_kwargs=try_to_record_lm_kwargs, program=program
+        )
+        for prompt_completion_dict in trace_prompt_comletion_data:
+            if keep_data_keys:
+                prompt_completion_dict = {**data_dict, **prompt_completion_dict}
+            prompt_completion_data.append(prompt_completion_dict)
+    return prompt_completion_data
+
+def build_messages_from_trace(
+        trace: List[Dict],
+        exclude_demos: bool=False,
+        try_to_record_lm_kwargs: bool = False,
+        program: Program = None,
+    ) -> Dict[str, List[Dict[str, Any]]]:
+    messages = []
+    # If the program is provided, build the predictor index to name mapping
+    if program:
+        pred_ind_to_name = {
+            ind: name for ind, (name, _) in enumerate(program.named_predictors())
+        }
+
+    # Build the prompt-completion data
+
+    adapter = dspy.settings.adapter or dspy.ChatAdapter()
+    data = []
+
+    # TODO: Make sure that this works for multi-stage pipelines
+    for pred_ind, (pred, inputs, outputs) in enumerate(trace):
+        # Get the demos from the predictor if exclude_demos is False
+        demos = [] if exclude_demos else pred.demos
+        messages = adapter.format(pred.signature, demos, inputs)
+        messages.append(adapter.format_turn(signature=pred.signature, values=outputs, role="assistant", incomplete=False))
+        data.append(messages)
+
+    return data
+
+def dummy_metric(example, pred, trace=None, frac=1.0):
```

Comment:
let's delete this, and just check if metric function is None.

Reply:
Reverting